apstra/api_iba_widgets.go

```diff
@@ -44,271 +35,131 @@ type IbaWidgetData struct {
 		aggregationPeriod = &td
 	}
 
 	var timeSeriesDuration *time.Duration
-	if o.TimeSeriesDuration != nil {
-		td := time.Duration(float64(*o.AggregationPeriod) * float64(time.Second))
+	if raw.TimeSeriesDuration != nil {
+		td := time.Duration(float64(*raw.TimeSeriesDuration) * float64(time.Second))
 		timeSeriesDuration = &td
 	}
 
-	return &IbaWidget{
-		Id:        o.Id,
-		CreatedAt: created,
-		UpdatedAt: updated,
-		Data: &IbaWidgetData{
-			AggregationPeriod:  aggregationPeriod,
-			OrderBy:            o.OrderBy,
-			StageName:          o.StageName,
-			ShowContext:        o.ShowContext,
-			Description:        o.Description,
-			AnomalousOnly:      o.AnomalousOnly,
-			SpotlightMode:      o.SpotlightMode,
-			ProbeId:            ObjectId(o.ProbeId),
-			Label:              o.Label,
-			Filter:             o.Filter,
-			TimeSeriesDuration: timeSeriesDuration,
-			DataSource:         o.DataSource,
-			MaxItems:           o.MaxItems,
-			CombineGraphs:      o.CombineGraphs,
-			VisibleColumns:     o.VisibleColumns,
-			Type:               *widgetType,
-			UpdatedBy:          o.UpdatedBy,
-		},
-	}, nil
+	*i = IbaWidgetData{
+		AggregationPeriod:  aggregationPeriod,
+		OrderBy:            raw.OrderBy,
+		StageName:          raw.StageName,
+		ShowContext:        raw.ShowContext,
+		Description:        raw.Description,
+		AnomalousOnly:      raw.AnomalousOnly,
+		SpotlightMode:      raw.SpotlightMode,
+		ProbeId:            ObjectId(raw.ProbeId),
+		Label:              raw.Label,
+		Filter:             raw.Filter,
+		TimeSeriesDuration: timeSeriesDuration,
+		DataSource:         raw.DataSource,
+		MaxItems:           raw.MaxItems,
+		CombineGraphs:      raw.CombineGraphs,
+		VisibleColumns:     raw.VisibleColumns,
+		UpdatedBy:          raw.UpdatedBy,
+		WidgetType:         t,
+	}
+
+	return err
 }
 
-func (o *IbaWidgetData) raw() *rawIbaWidget {
+func (i *IbaWidgetData) MarshalJSON() ([]byte, error) {
 	var aggPeriod, timeSeriesDuration int
-	if o.AggregationPeriod != nil {
-		aggPeriod = int(o.AggregationPeriod.Seconds())
+	if i.AggregationPeriod != nil {
+		aggPeriod = int(i.AggregationPeriod.Seconds())
 	} else {
 		aggPeriod = 1
 	}
-	if o.TimeSeriesDuration != nil {
-		timeSeriesDuration = int(o.TimeSeriesDuration.Seconds())
+	if i.TimeSeriesDuration != nil {
+		timeSeriesDuration = int(i.TimeSeriesDuration.Seconds())
 	} else {
 		timeSeriesDuration = 1
 	}
-	return &rawIbaWidget{
+
+	raw := struct {
+		AggregationPeriod  *int     `json:"aggregation_period,omitempty"`
+		OrderBy            string   `json:"orderby,omitempty"`
+		StageName          string   `json:"stage_name,omitempty"`
+		ShowContext        bool     `json:"show_context,omitempty"`
+		Description        string   `json:"description,omitempty"`
+		AnomalousOnly      bool     `json:"anomalous_only,omitempty"`
+		SpotlightMode      bool     `json:"spotlight_mode,omitempty"`
+		ProbeId            string   `json:"probe_id"`
```

Comment:
Use `ObjectId` to avoid type conversion?

Reply:
Done